content/blog/2020-12-14-parallel-Java/index.md

```diff
@@ -0,0 +1,126 @@
+We can define a special array type with the restriction that an object reference value o assigned to cell n (where n is a natural number constant) of such an array has a runtime type that is parameterized by n. If accesses through cell n touch only region n (even by following a chain of references), then the accesses through different cells are guaranteed to be disjoint. Such an array is being called index-parameterized array. 
+
+To represent such arrays, the authors introduce two language constructs: <br/>
+(1) An array PRL element written [e], where e is an integer expression.<br/>
+(2) An index-parameterized array type that allows programmer to write the region and type of array cell e using the array RPL element [e]. For example, programmers can specify that cell e resides in region Root:[e] and has type C<Root:[e]><br/>
+
+### 3. Subarrays
+A familiar pattern for writing divide and conquer recursion is to partition an array into two or more disjoint pieces and give each array to a subtask. <br/>
+(1) DPJ provides a class DPJArray that wraps an ordinary Java array and provides a view into contiguous segment of it, parameterized by start position S and length L.<br/>
+(2) DPJ provides a class DPJPartition, representing an indexed collection of DPJArray objects, all of which point into mutually disjoint segments of the original array.<br/>
+(3) To support recursive computations, the authors extend the syntax of RPLs by allowing a final local variable z of class type to appear at the head of an RPL which ensures that different partitions get different regions.<br/>
+
+### 4. Commutativity Annotations
+(1) Classes may contain declarations of the form m commuteswith m’, where m and m’ are method names, indicating that any pair of invocations of the named methods may be safely done in parallel, regardless of the read and write effects of the methods.<br/>
+(2) To represent and check effects soundly in the present of commutativity annotations, the effect system provides a novel invocation effect of the form invokes m with E, which records that an invocation of method m occurred with underlying effects E.<br/>
+
+In summary, the authors formalized a subset of DPJ, called CoreDPJ as a sequential language. For more details about the syntax, static and dynamic semantics, please refer to the paper. 
+
+## Evaluation
+The authors extended Sun’s javac compiler which compiles DPJ into common Java source. And they built a runtime system for DPJ using the ForkJoinTask framework that will be added to the java.util.concurrent standard library in Java 1.7 [8]. The benchmark used for evaluation are parallel merge sort, two codes from the Java Grande parallel benchmark suite (a Monte Carlo financial simulation and IDEA encryption), the force computation from the Barnes-Hut n-body simulation, k-means clustering from the STAMP benchmarks, and a tree-based collision detection algorithm from a large, real-world open source game engine called JMonkey.

```

Comment:
Instead of using a footnote for [8], which is just a URL for a website, why not use a [Markdown link](http://gee.cs.oswego.edu/dl/concurrency-interest)?